presto-spark-launcher/src/main/java/com/facebook/presto/spark/launcher/PrestoSparkRunner.java

```diff
@@ -88,7 +91,136 @@ public void run(
             Optional<String> queryDataOutputLocation)
     {
         IPrestoSparkQueryExecutionFactory queryExecutionFactory = driverPrestoSparkService.getQueryExecutionFactory();
+        try {
+            execute(user,
+                    principal,
+                    extraCredentials,
+                    catalog,
+                    schema,
+                    source,
+                    userAgent,
+                    clientInfo,
+                    clientTags,
+                    sessionProperties,
+                    catalogSessionProperties,
+                    sqlText,
+                    sqlLocation,
+                    sqlFileHexHash,
+                    sqlFileSizeInBytes,
+                    traceToken,
+                    sparkQueueName,
+                    queryStatusInfoOutputLocation,
+                    queryDataOutputLocation,
+                    queryExecutionFactory,
+                    Optional.empty());
+        }
+        catch (PrestoSparkFailure failure) {
+            // Session object is not created yet. So, we need to directly access sessionProperties map
+            String disableBroadcastJoinOnOOM = sessionProperties.get("spark_disable_broadcast_join_on_oom");
+            if (disableBroadcastJoinOnOOM != null && disableBroadcastJoinOnOOM.equalsIgnoreCase("true") && failure.isBroadcastJoinOOM()) {
```

Comment:
Let's make these tests very generic. In the future, we may have more retry strategies. We should get the retry signal from `PrestoSparkFailure` directly instead of having string/session comparison scared around